text/0000-placement-traits.md

```diff
@@ -0,0 +1,312 @@
+- Feature Name: placement_traits
+- Start Date: 2015-12-21
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC amends RFC #0809 to reduce the number of traits involved, take
+allocators into account, and pin down the story on DST placement.
+
+## Traits
+
+There are now three traits:
+
+1. `Placer` -- Placement in syntax:  `let owner = PLACE <- value`.
+2. `Boxer` -- Box syntax: `let boxed: Box<_> = box value`.
+3. `Place` -- An "out" pointer.
+
+## Allocators
+
+Boxer's now make Places from Allocators. This means that any type implementing
+`Boxer` can be allocated with any Allocator using placement in syntax (see the
+detailed design for more).
+
+## DSTs
+
+`Boxer::make_place` and `Placer::make_place` are bounded by `Data: Sized` to
+future proof against DST placement.
+
+Furthermore, this RFC explicitly defines the guarantees of when/where
+`Placer::make_place` will be called.
+
+# Detailed design
+[design]: #detailed-design
+
+The new trait hierarchy is as follows:
+
+```rust
+/// Interface to implementations of  `PLACE <- EXPR`.
+///
+/// `PLACE <- EXPR` effectively desugars into:
+///
+/// ```rust,ignore
+/// let p = PLACE;
+/// let mut place = Placer::make_place(p);
+/// let raw_place = Place::pointer(&mut place);
+/// let value = EXPR;
+/// unsafe {
+///     std::ptr::write(raw_place, value);
+///     Place::finalize(place)
+/// }
```

Comment:
For inference to work, all of this (and the `box` desugaring below) should be nested expressions (no blocks and statements), either with some safe `write_and_finalize` function (harder to enforce optimizations like we have on `std::ptr::write`) or with a special HIR node that is lowered to the appropriate MIR/LLVM IR.